### ISSUE TITLE
[RLLib] Importing bandits without tensorflow installed causes an import error

### ISSUE THREAD

**Issue by egeonat** [ray-project/ray#31327]
### What happened + What you expected to happen

**The Bug**
When I attempt to run the unmodified RLLib example script `env_rendering_and_recording.py` it crashes because of an import error. When I investigated the reason, I found out that unlike other algorithms bandit.py attempts to import both torch and tensorflow implementation classes, preventing the use of the library without having both deep learning frameworks installed. 

**Expected Behavior**
I would expect bandits.py to import one of the torch or tensorflow implementations based on the framework chosen by the user, as is the case for other algorithms like ppo.py

**Info**
Below is the output I get when I run the command: `python rllib/examples/env_rendering_and_recording.py --framework torch`
```
2022-12-27 10:54:41,244 INFO worker.py:1538 -- Started a local Ray instance.
Traceback (most recent call last):
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/tuner.py", line 272, in fit
    return self._local_tuner.fit()
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/impl/tuner_internal.py", line 420, in fit
    analysis = self._fit_internal(trainable, param_space)
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/impl/tuner_internal.py", line 520, in _fit_internal
    **self._get_tune_run_arguments(trainable),
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/impl/tuner_internal.py", line 483, in _get_tune_run_arguments
    if is_function_trainable(trainable):
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/registry.py", line 64, in is_function_trainable
    trainable = get_trainable_cls(trainable)
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/registry.py", line 45, in get_trainable_cls
    validate_trainable(trainable_name)
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/registry.py", line 55, in validate_trainable
    _register_all()
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/rllib/__init__.py", line 39, in _register_all
    register_trainable(key, get_trainable_class_and_config()[0])
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/rllib/algorithms/registry.py", line 65, in _import_bandit_lints
    from ray.rllib.algorithms.bandit.bandit import BanditLinTS
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/rllib/algorithms/bandit/__init__.py", line 1, in <module>
    from ray.rllib.algorithms.bandit.bandit import (
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/rllib/algorithms/bandit/bandit.py", line 6, in <module>
    from ray.rllib.algorithms.bandit.bandit_tf_policy import BanditTFPolicy
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/rllib/algorithms/bandit/bandit_tf_policy.py", line 8, in <module>
    from ray.rllib.algorithms.bandit.bandit_tf_model import (
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/rllib/algorithms/bandit/bandit_tf_model.py", line 2, in <module>
    import tensorflow_probability as tfp
ModuleNotFoundError: No module named 'tensorflow_probability'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "rllib/examples/env_rendering_and_recording.py", line 131, in <module>
    run_config=air.RunConfig(stop=stop),
  File "/home/onat/miniconda3/envs/rllib/lib/python3.7/site-packages/ray/tune/tuner.py", line 278, in fit
    ) from e
ray.tune.error.TuneError: The Ray Tune run failed. Please inspect the previous error messages for a cause. After fixing the issue, you can restart the run from scratch or continue this run. To continue this run, you can use `tuner = Tuner.restore("/home/onat/ray_results/PPO")`.
```

### Versions / Dependencies

Ray-2.2.0, Python 3.7.15, Ubuntu 20.04
Pytorch 11

### Reproduction script

conda create -n rllib python=3.7
conda activate rllib
pip install ray[rllib]
conda install pytorch==1.11.0 torchvision==0.12.0 torchaudio==0.11.0 cudatoolkit=11.3 -c pytorch
python rllib/examples/env_rendering_and_recording.py

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by ArturNiederfahrenhorst** [comment]
Hi @egeonat ,

Thanks for opening this issue. RLlib does not come with torch/TF, because we don't want to impose this dependency on users that only use one of the trameworks. Please install TF and TF probability if you want to use them for now.
I'm opening a PR that deals with these imports in a more graceful way because I've seen this cause confusion multiple times now 🙂

**Comment by ArturNiederfahrenhorst** [comment]
Fixed, thanks!

**Comment by axelbr** [comment]
It seems that this still persists in ray 2.3.1.

**Comment by mbusch-regis** [comment]
I think maybe @ArturNiederfahrenhorst misunderstands the question. I am using the examples but substituting my own environment. When I try to use tune.tuner, I get the same error as above, even though I explicitly tell it config.framework("torch").

```
  ppo_config = PPOConfig()
  print(ppo_config.clip_param)  
  ppo_config.training(lr=tune.grid_search([0.001, 0.0001]), clip_param=0.2)

  ppo_config.framework("torch")   # <<----- 

  ppo_config = ppo_config.environment(env=Pulse,  env_config={'config_insult':config_insult, 'episode_length':600, 'max_episodes':20, 'time_step':1})   
  tune.Tuner(  
      "PPO",
      run_config=air.RunConfig(stop={"episode_reward_mean": 200}),
      param_space=ppo_config.to_dict(),
  ).fit()
```

---

```
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[7], line 20
     13 ppo_config = ppo_config.environment(env=Pulse,  env_config={'config_insult':config_insult, 'episode_length':600, 'max_episodes':20, 'time_step':1})   
     14 # Use to_dict() to get the old-style python config dict
     15 # when running with tune.
     16 tune.Tuner(  
     17     "PPO",
     18     run_config=air.RunConfig(stop={"episode_reward_mean": 200}),
     19     param_space=ppo_config.to_dict(),
---> 20 ).fit()

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/tuner.py:292, in Tuner.fit(self)
    290 if not self._is_ray_client:
    291     try:
--> 292         return self._local_tuner.fit()
    293     except TuneError as e:
    294         raise TuneError(
    295             _TUNER_FAILED_MSG.format(
    296                 path=self._local_tuner.get_experiment_checkpoint_dir()
    297             )
    298         ) from e

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/impl/tuner_internal.py:455, in TunerInternal.fit(self)
    453 if not self._is_restored:
    454     param_space = copy.deepcopy(self._param_space)
--> 455     analysis = self._fit_internal(trainable, param_space)
    456 else:
    457     analysis = self._fit_resume(trainable)

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/impl/tuner_internal.py:560, in TunerInternal._fit_internal(self, trainable, param_space)
    555 def _fit_internal(
    556     self, trainable: TrainableType, param_space
    557 ) -> ExperimentAnalysis:
    558     """Fitting for a fresh Tuner."""
    559     args = {
--> 560         **self._get_tune_run_arguments(trainable),
    561         **dict(
    562             run_or_experiment=trainable,
    563             config={**param_space},
    564             num_samples=self._tune_config.num_samples,
    565             search_alg=self._tune_config.search_alg,
    566             scheduler=self._tune_config.scheduler,
    567             name=self._run_config.name,
    568             log_to_file=self._run_config.log_to_file,
    569         ),
    570         **self._tuner_kwargs,
    571     }
    572     analysis = run(
    573         **args,
    574     )
    575     self.clear_remote_string_queue()

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/impl/tuner_internal.py:523, in TunerInternal._get_tune_run_arguments(self, trainable)
    518         checkpoint_at_end = False
    519     # If this is a user-defined trainable, just keep the value
    520     # Function trainables will raise a runtime error later if set to True
    521 else:
    522     # Set default to False for function trainables and True for everything else
--> 523     if is_function_trainable(trainable):
    524         checkpoint_at_end = False
    525     else:

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/registry.py:68, in is_function_trainable(trainable)
     65 from ray.tune.trainable import FunctionTrainable
     67 if isinstance(trainable, str):
---> 68     trainable = get_trainable_cls(trainable)
     70 is_wrapped_func = isinstance(trainable, type) and issubclass(
     71     trainable, FunctionTrainable
     72 )
     73 return is_wrapped_func or (
     74     not isinstance(trainable, type)
     75     and (
   (...)
     79     )
     80 )

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/registry.py:45, in get_trainable_cls(trainable_name)
     43 @DeveloperAPI
     44 def get_trainable_cls(trainable_name):
---> 45     validate_trainable(trainable_name)
     46     return _global_registry.get(TRAINABLE_CLASS, trainable_name)

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/tune/registry.py:55, in validate_trainable(trainable_name)
     51 if not _has_trainable(trainable_name):
     52     # Make sure everything rllib-related is registered.
     53     from ray.rllib import _register_all
---> 55     _register_all()
     56     if not _has_trainable(trainable_name):
     57         raise TuneError("Unknown trainable: " + trainable_name)

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/rllib/__init__.py:35, in _register_all()
     32 from ray.rllib.algorithms.registry import ALGORITHMS, _get_algorithm_class
     34 for key, get_trainable_class_and_config in ALGORITHMS.items():
---> 35     register_trainable(key, get_trainable_class_and_config()[0])
     37 for key in ["__fake", "__sigmoid_fake_data", "__parameter_tuning"]:
     38     register_trainable(key, _get_algorithm_class(key))

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/rllib/algorithms/registry.py:64, in _import_bandit_lints()
     63 def _import_bandit_lints():
---> 64     from ray.rllib.algorithms.bandit.bandit import BanditLinTS
     66     return BanditLinTS, BanditLinTS.get_default_config()

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/rllib/algorithms/bandit/__init__.py:1
----> 1 from ray.rllib.algorithms.bandit.bandit import (
      2     BanditLinTS,
      3     BanditLinUCB,
      4     BanditLinTSConfig,
      5     BanditLinUCBConfig,
      6 )
      8 __all__ = [
      9     "BanditLinTS",
     10     "BanditLinUCB",
     11     "BanditLinTSConfig",
     12     "BanditLinUCBConfig",
     13 ]

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/rllib/algorithms/bandit/bandit.py:6
      4 from ray.rllib.algorithms.algorithm import Algorithm
      5 from ray.rllib.algorithms.algorithm_config import AlgorithmConfig
----> 6 from ray.rllib.algorithms.bandit.bandit_tf_policy import BanditTFPolicy
      7 from ray.rllib.algorithms.bandit.bandit_torch_policy import BanditTorchPolicy
      8 from ray.rllib.policy.policy import Policy

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/rllib/algorithms/bandit/bandit_tf_policy.py:8
      6 from gymnasium import spaces
      7 import ray
----> 8 from ray.rllib.algorithms.bandit.bandit_tf_model import (
      9     DiscreteLinearModelThompsonSampling,
     10     DiscreteLinearModelUCB,
     11     DiscreteLinearModel,
     12     ParametricLinearModelThompsonSampling,
     13     ParametricLinearModelUCB,
     14 )
     15 from ray.rllib.models.catalog import ModelCatalog
     16 from ray.rllib.models.modelv2 import restore_original_dimensions

File ~/anaconda3/envs/torch/lib/python3.10/site-packages/ray/rllib/algorithms/bandit/bandit_tf_model.py:2
      1 import gymnasium as gym
----> 2 import tensorflow_probability as tfp
      4 from ray.rllib.models.modelv2 import ModelV2
      5 from ray.rllib.models.tf.tf_modelv2 import TFModelV2

ModuleNotFoundError: No module named 'tensorflow_probability'
```

I'm not using TF, I don't want TF, but it seems to be hardcoded into your bandit.py file.

...Or, I'm just a noob who doesn't understand the code yet.

**Comment by Rohan138** [comment]
Thank you for bringing this up! Should be fixed by #34756
